pkg/auth/ldaputil/testclient/testclient.go

```diff
@@ -0,0 +1,133 @@
+	return
+}
+
+// StartTLS begins a TLS-wrapped LDAP connection
+func (c *Fake) StartTLS(config *tls.Config) error {
+	return nil
+}
+
+// Close closes an LDAP connection
+func (c *Fake) Close() {
+	return
+}
+
+// Bind binds to the LDAP server with a bind DN and password
+func (c *Fake) Bind(username, password string) error {
+	return nil
+}
+
+// SimpleBind binds to the LDAP server using the Simple Bind mechanism
+func (c *Fake) SimpleBind(simpleBindRequest *ldap.SimpleBindRequest) (*ldap.SimpleBindResult, error) {
+	return c.simpleBindResponse, nil
+}
+
+// Modify forwards a modification request to the LDAP server
+func (c *Fake) Modify(modifyRequest *ldap.ModifyRequest) error {
+	return nil
+}
+
+// Compare ... ?
+func (c *Fake) Compare(dn, attribute, value string) (bool, error) {
+	return false, nil
+}
+
+// PasswordModify forwards a password modify request to the LDAP server
+func (c *Fake) PasswordModify(passwordModifyRequest *ldap.PasswordModifyRequest) (*ldap.PasswordModifyResult, error) {
+	return c.passwordModifyResponse, nil
+}
+
+// Search forwards a search request to the LDAP server
+func (c *Fake) Search(searchRequest *ldap.SearchRequest) (*ldap.SearchResult, error) {
+	return c.searchResponse, nil
+}
+
+// SearchWithPaging forwards a search request to the LDAP server and pages the response
+func (c *Fake) SearchWithPaging(searchRequest *ldap.SearchRequest, pagingSize uint32) (*ldap.SearchResult, error) {
+	return c.searchResponse, nil
+}
+
+// NewMatchingSearchErrorClient returns a new matchingSeachError client sitting on top of the parent
+// client. This client returns the gien error when a search base DN matches the given base DN, and
+// defers to the parent otherwise.
+func NewMatchingSearchErrorClient(parent ldapclient.Client, baseDN string, returnErr error) ldapclient.Client {
+	return &matchingSearchErrClient{
+		Client:    parent,
+		baseDN:    baseDN,
+		returnErr: returnErr,
+	}
+}
+
+// matchingSearchErrClient returns the returnErr on every Search() where the search base DN matches the given DN
+// or defers the search to the parent client
+type matchingSearchErrClient struct {
+	ldapclient.Client
+	baseDN    string
+	returnErr error
+}
+
+func (c *matchingSearchErrClient) Search(searchRequest *ldap.SearchRequest) (*ldap.SearchResult, error) {
+	if searchRequest.BaseDN == c.baseDN {
+		return nil, c.returnErr
+	}
+	return c.Client.Search(searchRequest)
+}
+
+// NewDNMappingClient returns a new dnMappingClient sitting on top of the parent client. This client returns the
+// ldap entry mapped to with this DN in its' internal DN map, or defers to the parent if the DN is not mapped.
+func NewDNMappingClient(parent ldapclient.Client, dnMapping map[string]*ldap.Entry) ldapclient.Client {
+	return &dnMappingClient{
+		Client:    parent,
+		dnMapping: dnMapping,
+	}
+}
+
+// dnMappingClient returns the LDAP entry mapped to by the base dn given, or if no mapping happens, defers to the parent
+type dnMappingClient struct {
```

Comment:
I never ran into any cases ... it's like `PrependReactor`, just create the correct one and stick them together.